lib/core/src/Cardano/Wallet/Primitive/AddressDiscovery/Shared.hs

```diff
@@ -0,0 +1,313 @@
+-- CIP-XXX Multi-signature Wallets.
+--
+-- It indicates that the subtree of this node is used according to this
+-- specification.
+--
+-- Hardened derivation is used at this level.
+purposeCIPXXX :: Index 'Hardened 'PurposeK
+purposeCIPXXX = toEnum 0x8000002D
+
+-- | Create a SharedState from the all needed ingredients.
+-- There is no validation and it is unsafe way.
+unsafeSharedState
+    :: forall (n :: NetworkDiscriminant) k. k 'AccountK XPub
+    -> Index 'Hardened 'AccountK
+    -> AddressPoolGap
+    -> ScriptTemplate
+    -> Maybe ScriptTemplate
+    -> Map Address (Index 'Soft 'ScriptK, AddressState)
+    -> SharedState n k
+unsafeSharedState accXPub accIx g pTemplate dTemplate ourAdresesses =
+    SharedState
+    { shareStateAccountKey = accXPub
+    , shareStateDerivationPrefix = DerivationPrefix ( purposeCIPXXX, coinTypeAda, accIx )
+    , shareStateGap = g
+    , shareStatePaymentTemplate = pTemplate
+    , shareStateDelegationTemplate = dTemplate
+    , shareStateOurAddresses = ourAdresesses
+    }
+
+-- | Create a new SharedState.
+newSharedState
+    :: forall (n :: NetworkDiscriminant) k. k 'AccountK XPub
+    -> Index 'Hardened 'AccountK
+    -> AddressPoolGap
+    -> ScriptTemplate
+    -> Maybe ScriptTemplate
+    -> SharedState n k
+newSharedState accXPub accIx g pTemplate dTemplate =
+    extendSharedState (toEnum 0) $
+    unsafeSharedState accXPub accIx g pTemplate dTemplate Map.empty
+
+replaceCosignersWithVerKeys
+    :: ScriptTemplate
+    -> Int
+    -> Script KeyHash
+replaceCosignersWithVerKeys (ScriptTemplate xpubs scriptTemplate) index =
+    replaceCosigner scriptTemplate
+  where
+    replaceCosigner :: Script Cosigner -> Script KeyHash
+    replaceCosigner = \case
+        RequireSignatureOf c -> RequireSignatureOf $ toKeyHash c
+        RequireAllOf xs      -> RequireAllOf (map replaceCosigner xs)
+        RequireAnyOf xs      -> RequireAnyOf (map replaceCosigner xs)
+        RequireSomeOf m xs   -> RequireSomeOf m (map replaceCosigner xs)
+        ActiveFromSlot s     -> ActiveFromSlot s
+        ActiveUntilSlot s    -> ActiveUntilSlot s
+    toKeyHash :: Cosigner -> KeyHash
+    toKeyHash c =
+        let ix = toEnum index
+            (Just accXPub) = liftXPub <$> Map.lookup c xpubs
+            verKey = deriveMultisigPublicKey accXPub ix
+        in hashKey verKey
+
+constructAddressFromIx
+    :: CA.NetworkTag
+    -> ScriptTemplate
+    -> Maybe ScriptTemplate
+    -> Index 'Soft 'ScriptK
+    -> Address
+constructAddressFromIx tag pTemplate dTemplate ix =
+    let delegationCredential = DelegationFromScript . toScriptHash
+        paymentCredential = PaymentFromScript . toScriptHash
+        createAddress pScript' dScript' =
+            CA.unAddress $
+            delegationAddress tag
+            (paymentCredential pScript') (delegationCredential dScript')
+        dTemplate' = fromMaybe pTemplate dTemplate
```

Comment:
This sounds incorrect. If there's no delegation template, then the address should not include any delegation part. Defaulting to using payment credentials as delegation credentials should be done in the API, not here.

Reply:
ok, so this should be decided on client side - done https://github.com/input-output-hk/cardano-wallet/pull/2508/commits/9522d0c0a5b0fd39ffab1b03235f364c78a972f3